src/components/data/DataTable.js

```diff
@@ -114,6 +115,18 @@ const DataTable = props => {
           { billingProject: googleProject, dataReference: snapshotName } :
           { filterTerms: activeTextFilter })
       }))
+    // find all the unique attribute names contained in the current page of results
+    const attrNamesFromResults = _.uniq(_.flatMap(_.keys, _.map('attributes', results)))
+    // add any attribute names from the current page of results to those found in metadata.
+    // This allows for the stale metadata (e.g. the metadata cache is out of date).
+    // For the time being, the uniquness check MUST be case-insensitive (e.g. { sensitivity: 'accent' })
+    // in order to prevent case-divergent columns from being displayed, as that would expose some other bugs
+    const newAttrsForThisType = concatenateAttributeNames(entityMetadata[entityType]?.attributeNames, attrNamesFromResults)
+    if (newAttrsForThisType !== entityMetadata[entityType].attributeNames) {
+      const newMetadata = _.cloneDeep(entityMetadata)
+      newMetadata[entityType].attributeNames = newAttrsForThisType
+      setEntityMetadata(newMetadata)
+    }
```

Comment:
question for the UI experts: do I need to explicitly call `setEntityMetadata()` to persist my changes to state, or could I just use:
```entityMetadata[entityType].attributeNames = newAttrsForThisType```
both seem to function the same as I click through the app, but I don't know if there is an underlying difference.

Reply:
From the react FAQ, I see them doing this: `      setState(state => ({ ...state, left: e.pageX, top: e.pageY }));`, which has me voting for using the setter. The only actual difference I can see is that react has a mechanism to look at previous state compared to current state, and if we mutated current state than that wouldn't work, which would be confusing if someone was trying to use that mechanism. That being said, we aren't using that mechanism :)

Reply:
This is the way. Something like:
```javascript
setEntityMetadata(_.update(['attributeNames', entityType], newAttrsForThisType))
```
Note that Lodash's auto-currying is aiding concision here.